### ISSUE TITLE
I cant run my project on my server for testing proposes because 'Shrink-ray' is not installed because 'node-gyp' fails

### ISSUE THREAD

**Issue by mokingguy** [angular-fullstack/generator-angular-fullstack#2190]
- [x] I understand that GitHub issues are not for tech support, but for questions specific to this generator, bug reports, and feature requests.

Item | Version
----- | -----
generator-angular-fullstack | 4.0.1
Node | 6.2.0
npm | 3.10.6
Operating System | ~~CentOS~~ RedHat server
Python | 2.6 for system use but separately i have 2.7.12

Item | Answer
----- | -----
Transpiler | Babel
Markup | HTML 
CSS | CSS
Router | ngRoute 
Client Tests | Mocha
DB | MongoDB 
Auth | Y

The title kinda explains, i've seen the related questions but to no success.
The generator error is this.
```
Error: Cannot find module 'shrink-ray'
    at Function.Module._resolveFilename (module.js:440:15)
    at Function.Module._load (module.js:388:25)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/tararaapp-backend-panel/server/config/express.js:10:1)
    at Module._compile (module.js:541:32)
    at loader (/home/tararaapp-backend-panel/node_modules/babel-register/lib/node.js:146:5)
    at Object.require.extensions.(anonymous function) [as .js] (/home/tararaapp-backend-panel/node_modules/babel-register/lib/node.j
s:156:7)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/tararaapp-backend-panel/server/app.js:31:1)
    at Module._compile (module.js:541:32)
    at loader (/home/tararaapp-backend-panel/node_modules/babel-register/lib/node.js:146:5)
    at Object.require.extensions.(anonymous function) [as .js] (/home/tararaapp-backend-panel/node_modules/babel-register/lib/node.j
s:156:7)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
```
The npm installer tells me 
```
npm ERR! Failed at the iltorb@1.0.12 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the iltorb package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs iltorb
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls iltorb
npm ERR! There is likely additional logging output above.
```
I've tried to install node-gyp and it seems that it "worked" I have the requirements, except that when i use the command to tell node-gyp to use another python version.
`node-gyp --python /path/to/python2.7`
It just tells me how to use the command which isn't that useful

i've told npm to use the new python with 

`npm config set python /path/to/executable/python2.7`

i've used `npm i -g npm` to get the latest version of npm which for some people solved their problems.

I don't know what to do honestly. Any suggestions?

**Comment by Azayzel** [comment]
I know nothing about CentOS, with that being said, you still need build tools installed. In your case, CentOS being Linux, open terminal as root and run

# yum groupinstall 'Development Tools'

**Comment by GameHex** [comment]
This issue was already filed [here](https://github.com/angular-fullstack/generator-angular-fullstack/issues/2166). I ran into this problem and this solved it for me: 

Run `npm i -g npm`. There's a problem with one of the not newest npm versions not doing node-gyp correctly for shrink-ray. I just fixed it by updating npm, deleting the entire node_modules folder contents, then running npm install again.

Good luck!

**Comment by mokingguy** [comment]
@GameHex If you read my post you can see i've already tried that and that didn't work :(

**Comment by Awk34** [comment]
@mokingguy what does `npm --version` spit out? If it's not `3.10.6`, make sure to update it.

**Comment by mokingguy** [comment]
![screen shot 2016-08-31 at 1 46 03 p m](https://cloud.githubusercontent.com/assets/8469633/18142010/515b20dc-6f81-11e6-88b0-5f2d42ccf243.png)

**Comment by GameHex** [comment]
@mokingguy Ah, so you did! Try the other steps as well. I spent a good night trying to figure that problem out. Updating npm worked for me, but I also had to clean out the entire node_modules folder manually by moving the whole thing to the trash. Try to run `npm cache clean` as well as run `npm rebuild`. Some combination of that ended up working.

**Comment by Awk34** [comment]
@mokingguy you could try manually `node-gyp` building iltorb to see if you get a more useful error message.

**Comment by mokingguy** [comment]
@Awk34 I'm not sure how to do that

**Comment by mokingguy** [comment]
@GameHex I've done `npm cache clean` and `npm rebuild`, deleted the whole node_modules directory then done `npm install` and it gives me the same result

**Comment by GameHex** [comment]
@mokingguy Interesting! I can only imagine the problem lies with your node-gyp, then. There are more detailed installation steps and info on configuration [here](https://github.com/nodejs/node-gyp#installation). Perhaps something was missed your first time around trying to install it.

**Comment by Awk34** [comment]
@mokingguy `$ make` and `$ gcc` commands don't fail, correct?

You can use the [How to Use](https://github.com/nodejs/node-gyp#how-to-use) section of the node-gyp readme to see how to manually build an addon.

**Comment by mokingguy** [comment]
@GameHex  Perhaps, the only hiccup that i had is when i told node-gyp when the python 2.7 is, CentOS has python 2.6 installed but i manually installed 2.7 at /usr/src/Python-2.7.12 so when I used the command that they told me to use:
`node-gyp --python /usr/src/Python-2.7.12`
It spat back at me this
``` 
Usage: node-gyp <command> [options]                                                                                               
                                                                                                                                    
  where <command> is one of:                                                                                                        
    - build - Invokes `make` and builds the module                                                                                  
    - clean - Removes any generated build files and the "out" dir                                                                   
    - configure - Generates a Makefile for the current module                                                                       
    - rebuild - Runs "clean", "configure" and "build" all at once                                                                   
    - install - Install node development files for the specified node version.                                                      
    - list - Prints a listing of the currently installed node development files                                                     
    - remove - Removes the node development files for the specified version

node-gyp@3.4.0  /usr/local/lib/node_modules/node-gyp
node@6.2.0
```

witch doesn't really tells me anything  useful, did it do it, did it not? Idk

Everything else worked, I have make and gcc installed correctly.

![screen shot 2016-08-31 at 2 26 52 p m](https://cloud.githubusercontent.com/assets/8469633/18143285/0488b610-6f87-11e6-9fd8-f327045959b9.png)

 @Awk34 I'm going to read that and report back

**Comment by mokingguy** [comment]
I'm so sorry seems that I have redhat not CentOS, i don't know if it makes a diference or not...

**Comment by Awk34** [comment]
@mokingguy I believe the usage is `node-gyp --python /path/ <command>`, like `node-gyp --python /path/ build` from the addon's root directory

**Comment by mokingguy** [comment]
@Awk34 I think i did it, correct me if i'm wrong, I've cloned the litorb repo in my server used 
`node-gyp configure` & `node-gyp build`and it spat back this

```
root@server [/home/tararaapp-backend-panel/litorbTest]# git clone https://github.com/MayhemYDG/iltorb.git                           
Cloning into 'iltorb'...
remote: Counting objects: 344, done.                                                                                                
remote: Total 344 (delta 0), reused 0 (delta 0), pack-reused 344                                                                    
Receiving objects: 100% (344/344), 21.78 MiB | 8.37 MiB/s, done.
Resolving deltas: 100% (195/195), done.
Checking connectivity... done.

root@server [/home/tararaapp-backend-panel/litorbTest]# node-gyp configure                                                          
gyp info it worked if it ends with ok
gyp info using node-gyp@3.4.0
gyp info using node@6.2.0 | linux | x64
gyp info spawn /usr/bin/python2
gyp info spawn args [ '/usr/local/lib/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/home/tararaapp-backend-panel/litorbTest/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/local/lib/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/root/.node-gyp/6.2.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/root/.node-gyp/6.2.0',
gyp info spawn args   '-Dnode_gyp_dir=/usr/local/lib/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=node.lib',
gyp info spawn args   '-Dmodule_root_dir=/home/tararaapp-backend-panel/litorbTest',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
root@server [/home/tararaapp-backend-panel/litorbTest]# node-gyp configure                                                      
gyp info it worked if it ends with ok                                                                                               
gyp info using node-gyp@3.4.0                                                                                                       
gyp info using node@6.2.0 | linux | x64                                                                                             
gyp info spawn /usr/bin/python2                                                                                                     
gyp info spawn args [ '/usr/local/lib/node_modules/node-gyp/gyp/gyp_main.py',                                                       
gyp info spawn args   'binding.gyp',                                                                                                
gyp info spawn args   '-f',                                                                                                         
gyp info spawn args   'make',                                                                                                       
gyp info spawn args   '-I',                                                                                                         
gyp info spawn args   '/home/tararaapp-backend-panel/litorbTest/build/config.gypi',                                                 
gyp info spawn args   '-I',                                                                                                         
gyp info spawn args   '/usr/local/lib/node_modules/node-gyp/addon.gypi',                                                            
gyp info spawn args   '-I',                                                                                                         
gyp info spawn args   '/root/.node-gyp/6.2.0/include/node/common.gypi',                                                             
gyp info spawn args   '-Dlibrary=shared_library',                                                                                   
gyp info spawn args   '-Dvisibility=default',                                                                                       
gyp info spawn args   '-Dnode_root_dir=/root/.node-gyp/6.2.0',                                                                      
gyp info spawn args   '-Dnode_gyp_dir=/usr/local/lib/node_modules/node-gyp',                                                        
gyp info spawn args   '-Dnode_lib_file=node.lib',                                                                                   
gyp info spawn args   '-Dmodule_root_dir=/home/tararaapp-backend-panel/litorbTest',                                                 
gyp info spawn args   '--depth=.',                                                                                                  
gyp info spawn args   '--no-parallel',                                                                                              
gyp info spawn args   '--generator-output',                                                                                         
gyp info spawn args   'build',                                                                                                      
gyp info spawn args   '-Goutput_dir=.' ]                                                                                            
gyp: binding.gyp not found (cwd: /home/tararaapp-backend-panel/litorbTest) while trying to load binding.gyp                         
gyp ERR! configure error                                                                                                            
gyp ERR! stack Error: `gyp` failed with exit code: 1                                                                                
gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node_modules/node-gyp/lib/configure.js:305:16)                          
gyp ERR! stack     at emitTwo (events.js:106:13)                                                                                    
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)                                                                           
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:204:12)                                        
gyp ERR! System Linux 2.6.32-642.1.1.el6.x86_64                                                                                     
gyp ERR! command "/usr/local/bin/node" "/usr/local/bin/node-gyp" "configure"                                                        
gyp ERR! cwd /home/tararaapp-backend-panel/litorbTest                                                                               
gyp ERR! node -v v6.2.0                                                                                                             
gyp ERR! node-gyp -v v3.4.0                                                                                                         
gyp ERR! not ok            
                                                                                                         
root@server [/home/tararaapp-backend-panel/litorbTest]# node-gyp build                                                              
gyp info it worked if it ends with ok
gyp info using node-gyp@3.4.0
gyp info using node@6.2.0 | linux | x64
gyp info spawn make
gyp info spawn argsmake: Entering directory `/home/tararaapp-backend-panel/litorbTest/build'
make: *** No targets specified and no makefile found.  Stop.
make: Leaving directory `/home/tararaapp-backend-panel/litorbTest/build'
 [ 'BUILDTYPE=Release', '-C', 'build' ]
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:204:12)
gyp ERR! System Linux 2.6.32-642.1.1.el6.x86_64
gyp ERR! command "/usr/local/bin/node" "/usr/local/bin/node-gyp" "build"
gyp ERR! cwd /home/tararaapp-backend-panel/litorbTest
gyp ERR! node -v v6.2.0
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok
```

So i have to tell node-gyp where is python every time i have to use it? that's dumb tbh

**Comment by mokingguy** [comment]
I feel so dumb, i was the wrong directory. Here is the node-gyp error inside iltorb directory...

```
root@server [/home/tararaapp-backend-panel/litorbTest/iltorb]# node-gyp --python /usr/src/Python-2.7.12/python configure            
gyp info it worked if it ends with ok
gyp info using node-gyp@3.4.0
gyp info using node@6.2.0 | linux | x64
gyp info spawn /usr/src/Python-2.7.12/python
gyp info spawn args [ '/usr/local/lib/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/home/tararaapp-backend-panel/litorbTest/iltorb/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/local/lib/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/root/.node-gyp/6.2.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/root/.node-gyp/6.2.0',
gyp info spawn args   '-Dnode_gyp_dir=/usr/local/lib/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=node.lib',
gyp info spawn args   '-Dmodule_root_dir=/home/tararaapp-backend-panel/litorbTest/iltorb',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
gyp info ok 

root@server [/home/tararaapp-backend-panel/litorbTest/iltorb]# node-gyp --python /usr/src/Python-2.7.12/python build                
gyp info it worked if it ends with ok
gyp info using node-gyp@3.4.0
gyp info using node@6.2.0 | linux | x64
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory `/home/tararaapp-backend-panel/litorbTest/iltorb/build'
make: *** No rule to make target `Release/obj.target/decode/brotli/common/dictionary.o', needed by `Release/obj.target/decode.node'.
  Stop.
make: Leaving directory `/home/tararaapp-backend-panel/litorbTest/iltorb/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:204:12)
gyp ERR! System Linux 2.6.32-642.1.1.el6.x86_64
gyp ERR! command "/usr/local/bin/node" "/usr/local/bin/node-gyp" "--python" "/usr/src/Python-2.7.12/python" "build"
gyp ERR! cwd /home/tararaapp-backend-panel/litorbTest/iltorb
gyp ERR! node -v v6.2.0
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok 
```

**Comment by Azayzel** [comment]
Set your PYTHONPATH environment variable and you shouldn't have a problem

**Comment by mokingguy** [comment]
@Azayzel I've read somewhere that if i modify the python that the system uses it can break some stuff, specifically i've seen that it can break yum. I don't really know but i've got some stuff on that server running and I wouldn't risk it.

**Comment by Azayzel** [comment]
Is 'Release/obj.target/decode/brotli/common/dictionary.o' a valid file path? Does that file exist in that directory?

**Comment by mokingguy** [comment]
@Azayzel Now that you mention it, it does not. But i literally copied the source code from iltorb and it fails to build with node-gyp.

@Awk34 It seems that this an issue with iltorb, should I post this issue there and close this here, because it seems that does not have to do anything to do with the generator itself, or maybe you guys can help me solve it but I understand that is not your fault nor your issue.

**Comment by Awk34** [comment]
@mokingguy make an issue there, and link to this one.

**Comment by mokingguy** [comment]
I've solved it, you need gcc v4.8. Commenting here if someone tumbles in the same error.

Thank you guys so much you're awesome!.
